src/Elastic.Apm/Metrics/MetricsProvider/ProcessTotalCpuTimeProvider.cs

```diff
@@ -0,0 +1,61 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Elastic.Apm.Api;
+
+namespace Elastic.Apm.Metrics.MetricsProvider
+{
+	public class ProcessTotalCpuTimeProvider : IMetricsProvider
+	{
+		private const string ProcessCpuTotalPct = "system.process.cpu.total.norm.pct";
+		private readonly object _lock = new object();
+		private bool _first = true;
+		private TimeSpan _lastCurrentProcessCpuTime;
+		private DateTime _lastTick;
+
+		private Version _processAssemblyVersion;
+
+		public int ConsecutiveNumberOfFailedReads { get; set; }
+		public string NameInLogs => "total process CPU time";
+
+		public IEnumerable<Sample> GetValue()
+		{
+			var timespan = DateTime.UtcNow;
+			var cpuUsage = Process.GetCurrentProcess().TotalProcessorTime;
+			var currentTimeSpan = DateTime.UtcNow;
+
+			if (!_first)
+			{
+				if (_processAssemblyVersion == null)
+					_processAssemblyVersion = typeof(Process).Assembly.GetName().Version;
+
+				double cpuUsedMs;
+
+				//workaround for a CoreFx bug. See: https://github.com/dotnet/corefx/issues/37614#issuecomment-492489373
+				if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX) && _processAssemblyVersion < new Version(4, 3, 0))
+					cpuUsedMs = (cpuUsage - _lastCurrentProcessCpuTime).TotalMilliseconds / 100;
+				else
+					cpuUsedMs = (cpuUsage - _lastCurrentProcessCpuTime).TotalMilliseconds;
+
+				var totalMsPassed = (currentTimeSpan - _lastTick).TotalMilliseconds;
+				var cpuUsageTotal = cpuUsedMs / (Environment.ProcessorCount * totalMsPassed);
+
+
+				_first = false;
+				_lastTick = timespan;
+				_lastCurrentProcessCpuTime = cpuUsage;
+				return new List<Sample> { new Sample(ProcessCpuTotalPct, cpuUsageTotal) };
+			}
+
+			lock (_lock)
+			{
+				_first = false;
+				_lastTick = timespan;
+				_lastCurrentProcessCpuTime = cpuUsage;
+			}
+
+			return null;
```

Comment:
I would fine your proposal more complicated. With this there is just 1 method, which may or may not return the value - it almost always does, except the first time.

Reply:
Can you point to any .NET Framework API that uses this approach?

If the calls have completely different semantics using the same method for both is confusing. If you need to use "except for the first time" in a description I would say it's a strong sign to reconsider.

Reply:
I don't really have a perfect fitting example, but let's say take a look at the perf. counter API. `PerformanceCounter.NextValue` either returns a value, or throws an exception, you don't need to call a 'Prepare' method, or something like that - in our case we either return a value a value or `null` or empty enumeration.



> If the calls have completely different semantics using the same method for both is confusing

That's not the case. The semantics is the same: returns a value or null or an empty `IEnumerable`. 

The API design is the following: there is 1 method, named `GetSamples()` (renamed from `GetValue()`) and that returns `IEnumerable<MetricSample>`. This can be null or an empty `IEnumerable`, reason for that can be multiple things (not supported, the API we use returned nothing, or returned with error, etc.). I think this is a very simple design, and adding a preparation logic would make it more complicated - and most if the implementations don't need a preparation logic.

Yes, in case of the CPU time implementations the 1. call does not return a value, but we call this method periodically - that's the whole point of it. Why not take advantage of this and just call it and handle cases with `null` or empty `IEnumerable`? We need those checks anyway, because we can't be sure we always get values. Most of the `IMetricsProvider` providers don't need a `Prepare` step, so adding it to the interface or handling CPU metrics differently compared to other metrics, just because the 1. call does not return a value to me seems very complicated, it makes the code more complex without any real benefit.

Reply:
Turns out Python and Ruby do exactly the same.

But @axw had a very useful comment on this one (and that's what the Go agent does): we can capture the time current timestamp and cpu-time in the constructor and use that in the first call.

Implemented it, already pushed,it works, the code is shorter and we don't introduce any unnecessary abstraction. Love it. Strongly feel good about this. - this is about the original comment in this thread, remaining question/issues/wishes will be addressed soon.

Reply:
When I suggested to have a dedicated method (`Prepare`) I thought it's better than doing the first data read in constructor because duration between `Prepare` and the first timer callback is `MetricsInterval` while duration between constructor and the first timer callback might be much longer than `MetricsInterval` but it's very similar so I'm okay either way.